docs/src/design/chiplets/kernel_rom.md

```diff
@@ -1,91 +1,101 @@
+The variables $u_{init}$ and $u_{bus}$ represent reduced bus messages containing a kernel procedure hash.
+Denoting the random values received from the verifier as $\alpha_0, \alpha_1$, etc., this can be defined as
 
 $$
-v = \alpha_0 + \alpha_1 \cdot op_{krom} + \sum_{i=0}^3 (\alpha_{i + 2} \cdot r_i)
+\begin{aligned}
+u_{init} &= \alpha_0 + \sum_{i=0}^3 (\alpha_{i + 1} \cdot r_i) \\
+u_{bus} &= \alpha_0 + \alpha_1 \cdot op_{krom} + \sum_{i=0}^3 (\alpha_{i + 2} \cdot r_i) \\
+\end{aligned}
 $$
 
-Where, $op_{krom}$ is the unique [operation label](./main.md#operation-labels) of the kernel procedure call operation.
+Here, $op_{krom}$ is the unique [operation label](./main.md#operation-labels) of the kernel procedure call operation.
 
-The request side of the constraint for the operation is enforced during program block hashing of the [`SYSCALL` operation](../decoder/constraints.md#block-hash-computation-constraints).
+_Note: the above might be unsafe since the first element of a hash may collide with another bus message op code._
 
-To provide accessed kernel procedures to the chiplets bus, we must send the kernel procedure to the bus every time it is called, which is indicated by the $s_0$ column.
+To provide accessed kernel procedures to the chiplets bus, we must send the kernel procedure to the bus every time it is
+called, which is indicated by the $s_{first}$ column.
+We also ensure that each new hash in the table corresponds to one of the kernel procedure roots provided via public inputs.
+These responses are mutually exclusive, allowing us to group them in a single constraint.
 
 > $$
-b'_{chip} = b_{chip} \cdot (s_0 \cdot v + 1 - s_0) \text{ | degree} = 3
+b'_{chip} = b_{chip} \cdot (s_{first} \cdot u_{init} + (1 - s_{first}) \cdot u_{bus}) \text{ | degree} = 3
 $$
 
-Thus, when $s_0 = 0$ this reduces to $b'_{chip} = b_{chip}$, but when $s_0=1$ it becomes $b'_{chip} = b_{chip} \cdot u$.
-
-## Kernel procedure table constraints
-*Note: Although this table is described independently, it is implemented as part of the [chiplets virtual table](../chiplets/main.md#chiplets-virtual-table), which combines all virtual tables required by any of the chiplets into a single master table.*
-
-This kernel procedure table keeps track of all *unique* kernel function roots. The values in this table will be updated only when the value in the `idx` column changes.
-
-The row value included into $vt_{chip}$ is:
-
-$$
-v = \alpha_0 + \alpha_1 \cdot idx + \sum_{i=0}^3 (\alpha_{i + 2} \cdot r_i)
-$$
-
-The constraint against $vt_{chip}$ is:
-
-> $$
-vt_{chip}' = vt_{chip} \cdot (\Delta idx \cdot v + 1 - \Delta idx) \text{ | degree} = 3
-$$
+We also need to impose boundary constraints to ensure that the set of unique hashes in the trace matches the set of kernel procedure roots provided by the verifier.
+This is achieved by enforcing that the running product column for the bus is initialized with all $u_{init}$ hashes provided by the verifier.
```

Comment:
nit: "... with the `u_init` corresponding to each kernel procedure digest ..." or something similar